### PR TITLE
chore(internal): track config diff during startup

### PR DESCRIPTION
**Pull request by hikchoi**
# chore(internal): track config diff during startup
This PR:
- Adds new properties `dendronConfigChanged` and and `changedDendronConfigs`, which tracks configs that deviated from the default configuration defined

# Pull Request Checklist
### Basics

- [x] code should follow [Code Conventions](dev.process.code)
- [x] circular dependency check: make sure your code is not introducing new circular dependencies in plugin-core.  See [Avoiding Circular Dependencies](dev.process.code.best-practices).
- [x] sticking to existing conventions instead of creating new ones
  - eg: [if configuration for utilities are already in one module or package, add future utilities there as well](https://github.com/dendronhq/dendron/pull/1960#discussion_r786228021)

### Extended
- General
  - [x] check whether code be simplified
  - [x] check if similar function already exist in the codebase. if so, can it be re-used?
  - [x] check if this change adversely impact performance
- Operations
  - [x] when shipping this change, will it just work or will it introduce additional operational overhead due to complicated interface or known bugs?
- Architecture
  - [x] check if code is introducing changes on a foundational class or interface. if so, call for design review if needed
    - eg: [making changes to DNode](https://github.com/dendronhq/dendron/pull/2158#pullrequestreview-854689586)

### Basics
- [x] if you are adding analytics related changes, make sure the [Telemetry](https://wiki.dendron.so/notes/84df871b-9442-42fd-b4c3-0024e35b5f3c.html) docs are updated

### Extended
- [x] can we track the performance of this change to know if it is _successful_?
  - eg: [see usage for export pod](https://github.com/dendronhq/dendron/pull/2190#pullrequestreview-855715612)

### Basics

- [x] [Write Tests](dev.process.qa.test) 
- [x] [Confirm existing tests pass](dev.process.qa.test)
- [x] [Confirm manual testing](dev.process.qa.test) 
- [x] Common cases tested
- [x] 1-2 Edge cases tested
- [x] If your tests changes an existing snapshot, snapshots have been [updated](dev.process.qa.test)

### Extended
- [~] If you are adding a new language feature (graphically visible in VS Code/preview/publishing), an example is included in the [test workspace](dev.ref.test-workspace)
- CSS
  - [~] display is correct for following dimensions
    - [~] sm: screen ≥ 576px, eg. iphonex, (375x812)
    - [~] lg: screen ≥ 992px
    - [~] xxl: screen ≥ 1600px eg. mac (1600x900)
  - [~] display is correct for following browsers (across the various dimensions)
    - [~] safari
    - [~] firefox
    - [~] chrome

## Docs
- [~] if your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR
- [~] does this change introduce a new or better way of doing things that others need to be aware of? if so, an async should be created and a process added in [Development](dev) or [Packages](pkg)

### Extended
- [~]  is this a developer BREAKING change? if another person cloning from this branch will need to adjust their dependencies or mental model of the architecture, then it is. if this is the case, make sure this is communicated according to [Close Loop](dev.process.close-loop)

### CODE REVIEW

File: packages/engine-test-utils/src/__tests__/engine-server/configUtils.spec.ts [hunk 1]
```diff
@@ -9,6 +9,64 @@ function getConfig(
...
+      });
+    });
+    describe("WHEN v4 config is given", () => {
+      test("THEN return empty list", () => {
+        const config = ConfigUtils.genDefaultV4Config();
+        const output = ConfigUtils.findDifference({ config });
+        expect(output.length).toEqual(0);
+      });
+    });
+    describe("WHEN changed config is given", () => {
```
Comment (kevinslin) [inline packages/engine-test-utils/src/__tests__/engine-server/configUtils.spec.ts:27]: can we also add a test where the config only differes by vaults? we don't want to emit a diff in that case

File: packages/common-all/src/utils/index.ts [hunk 2]
```diff
@@ -1286,6 +1286,93 @@ export class ConfigUtils {
...
+   * The value is either a primitive value, or a stringified array.
+   *
+   * If comparing the array value of a config is unnecessary,
+   * make sure to add it to the omit path.
+   */
+  static flattenConfigObject(opts: { obj: Object; omitPaths?: string[] }) {
+    const { obj, omitPaths } = opts;
+    if (omitPaths && omitPaths.length > 0) {
+      omitPaths.forEach((path) => {
+        _.unset(obj, path);
```
Comment (kevinslin) [inline packages/common-all/src/utils/index.ts:1305]: isn't this mutating the original config? can we make a copy?

**Review by kevinslin — CHANGES_REQUESTED** [review]

File: packages/common-all/src/utils/index.ts (hunk 2 above)
Comment (hikchoi) [reply packages/common-all/src/utils/index.ts:1305]: right. good catch!

**Comment by hikchoi** [comment]
update:

- `InitializeWorkspace` is getting a little crowded, so adding a new event `ConfigEvents.ConfigChangeDetected`, which will track the actual diffs
- `InitializeWorkspace` will have `dendronConfigChanged` and `numConfigChanged` for ease of analysis.

`ConfigChangeDetected`'s payload will be a single stringified JSON:

e.g.)
```js
properties: {
    appVersion: '0.112.0',
    arch: 'arm64',
    changed: '[{"path":"workspace.journal.dailyDomain","value":"dailyz"},{"path":"workspace.task.taskCompleteStatus","value":"[\"x\",\"done\"]"},{"path":"workspace.enableFullHierarchyNoteTitle","value":true},{"path":"preview.enableNoteTitleForLink","value":false},{"path":"preview.enableHashesForFMTags","value":true}]',
    firstWeekSinceInstall: false,
    ideFlavor: 'Visual Studio Code',
    ideVersion: '1.70.2',
    nodeVersion: 'v16.13.2',
    userAgent: 'vscode',
    vscodeSessionId: redacted
  }
```

**Pull request merged**